binary_search_tree/tree.py

```diff
@@ -14,46 +17,187 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+            return self.root
+        else:
+            return self.add_helper(self.root, key, value)
+        
+    def add_helper(self, cur, key, value):
+        if cur == None:
+            return TreeNode(key, value)
+        else:
+            if key < cur.key:
+                cur.left = self.add_helper(cur.left, key, value)
+                return cur
+            else:
+                cur.right = self.add_helper(cur.right, key, value)
+                return cur
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
```

Comment:
👍  However space complexity is O(log n) if the tree is balanced and O(n) if not because of the recursive call stack.